### PR TITLE
Managed Disk Resource: Added support for network_access_policy

### PR DESCRIPTION
**Pull request by cmendible**
Fixes: #9796

### CODE REVIEW

**Comment by martinbjorgan** [comment]
@cmendible I was also working on this, you beat me to it :) 
I think we also should add the `disk_access_id` parameter to the managed disk resource since that is required if `AllowPrivate` is used?

**Comment by cmendible** [comment]
Hi! That's the reason I added the WIP prefix. Wwe need to be sure what else is needed for that specific case...

**Comment by martinbjorgan** [comment]
@cmendible Missed  that! Sorry :)

**Comment by cmendible** [comment]
@martinbjorgan confirmed [here](https://docs.microsoft.com/en-us/azure/virtual-machines/disks-enable-private-links-for-import-export-portal?WT.mc_id=AZ-MVP-5002618) that `disk_access_id` is required when the NeworkAccessPolicy is set to `AllowPrivate` and also that the field is supported in update operations: [Disks - Create Or Update](https://docs.microsoft.com/en-us/rest/api/compute/disks/createorupdate?WT.mc_id=AZ-MVP-5002618).

Can you open an issue to add `azurerm_disk_access` as a new resource so you can tackle that one?

**Comment by martinbjorgan** [comment]
@cmendible Sure, I'll give it a whirl

**Comment by cmendible** [comment]
PR #9889 (adds the `azurerm_disk_access` resource) must be merged so we can add a test for `network_access_policy = AllowPrivate` which requires a `disk_access_id`

**Comment by martinbjorgan** [comment]
@cmendible I'm almost done with tests for the resource, data source is done. 
Hopefully I'm done with those after the 24th

**Comment by cmendible** [comment]
> PR #9889 (adds the `azurerm_disk_access` resource) must be merged so we can add a test for `network_access_policy = AllowPrivate` which requires a `disk_access_id`

PR #9889 was merged and test for `network_access_policy = AllowPrivate` are in place.

@katbyte can you please review this one?

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 1]
```diff
@@ -322,6 +323,147 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+			ExpectError: acceptance.RequiresImportError("azurerm_managed_disk"),
+		},
+	})
+}
+
+// nolint unparam
+func testCheckManagedDiskExists(resourceName string, shouldExist bool) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		client := acceptance.AzureProvider.Meta().(*clients.Client).Compute.DisksClient
+		ctx := acceptance.AzureProvider.Meta().(*clients.Client).StopContext
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:438]: could we remove this and use `data.CheckWithClient`? as we are removing all usages of acceptance.AzureProvider for binary testing

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @cmendible - overall this looks great! i've just left some comments inline to address before merge

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @cmendible - but i think this change is going to break existing configs as tests are now failing with:

```
An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_managed_disk.test will be updated in-place
          ~ resource "azurerm_managed_disk" "test" {
                id                    = "/subscriptions/*******/resourceGroups/acctestRG-210415031958219440/providers/Microsoft.Compute/disks/acctestd-210415031958219440"
                name                  = "acctestd-210415031958219440"
              - network_access_policy = "AllowAll" -> null
                tags                  = {
                    "cost-center" = "ops"
                    "environment" = "acctest"
                }
                # (7 unchanged attributes hidden)
        
                # (1 unchanged block hidden)
            }
```

maybe we need to mark that as computed?

**Comment by cmendible** [comment]
> Thanks @cmendible - but i think this change is going to break existing configs as tests are now failing with:
> 
> ```
> An execution plan has been generated and is shown below.
>         Resource actions are indicated with the following symbols:
>           ~ update in-place
>         
>         Terraform will perform the following actions:
>         
>           # azurerm_managed_disk.test will be updated in-place
>           ~ resource "azurerm_managed_disk" "test" {
>                 id                    = "/subscriptions/*******/resourceGroups/acctestRG-210415031958219440/providers/Microsoft.Compute/disks/acctestd-210415031958219440"
>                 name                  = "acctestd-210415031958219440"
>               - network_access_policy = "AllowAll" -> null
>                 tags                  = {
>                     "cost-center" = "ops"
>                     "environment" = "acctest"
>                 }
>                 # (7 unchanged attributes hidden)
>         
>                 # (1 unchanged block hidden)
>             }
> ```
> 
> maybe we need to mark that as computed?

@katbyte I ran some tests and it should be fixed.

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 2]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
 	})
 }
 
+func TestAccAzureRMManagedDisk_create_withNetworkPolicy(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:325]: for consistency 
```suggestion
func TestAccAzureRMManagedDisk_networkPolicy(t *testing.T) {
```

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 3]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+		{
+			Config: testAccAzureRMManagedDisk_create_withNetworkPolicy(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+	})
+}
+
+func TestAccAzureRMManagedDisk_update_withNetworkPolicy(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:339]:
```suggestion
func TestAccAzureRMManagedDisk_networkPolicy_update(t *testing.T) {
```

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 4]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+			Config: testAccAzureRMManagedDisk_update_withNetworkPolicy(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				resource.TestCheckResourceAttr(data.ResourceName, "network_access_policy", "DenyAll"),
+			),
+		},
+	})
+}
+
+func TestAccAzureRMManagedDisk_import_withNetworkPolicy(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:361]: i don't think we need this as it duplicate the main import check?

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 5]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+	data := acceptance.BuildTestData(t, "azurerm_managed_disk", "test")
+	r := ManagedDiskResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: testAccAzureRMManagedDisk_create_withNetworkPolicy(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:335]: after every step in every test we generally have an import step, could we add these please?

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 6]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+		{
+			Config:      testAccAzureRMManagedDisk_import_withNetworkPolicy(data),
+			ExpectError: acceptance.RequiresImportError("azurerm_managed_disk"),
+		},
+	})
+}
+
+func TestAccAzureRMManagedDisk_create_withAllowPrivateNetworkPolicy(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:379]:
```suggestion
func TestAccAzureRMManagedDisk_allowPrivateNetworkPolicy(t *testing.T) {
```

ad so forth

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 7]
```diff
@@ -322,6 +322,117 @@ func TestAccManagedDisk_attachedStorageTypeUpdate(t *testing.T) {
...
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				resource.TestCheckResourceAttr(data.ResourceName, "network_access_policy", "AllowPrivate"),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccAzureRMManagedDisk_import_withAllowPrivateNetworkPolicy(t *testing.T) {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:418]: ditto here, i think we can drop this test

File: azurerm/internal/services/compute/managed_disk_resource_test.go [hunk 8]
```diff
@@ -1031,3 +1142,198 @@ resource "azurerm_linux_virtual_machine" "test" {
 }
 `, data.RandomInteger, data.Locations.Primary, data.RandomInteger, data.RandomInteger, data.RandomInteger)
 }
+
+func testAccAzureRMManagedDisk_create_withNetworkPolicy(data acceptance.TestData) string {
```
Comment (katbyte) [inline azurerm/internal/services/compute/managed_disk_resource_test.go:1146]: and then update the config function names

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @cmendible - overall looks good and i've left some comments inline

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @cmendible - aside from the tests failing with:
```
------- Stdout: -------
=== RUN   TestAccAzureRMManagedDisk_networkPolicy
=== PAUSE TestAccAzureRMManagedDisk_networkPolicy
=== CONT  TestAccAzureRMManagedDisk_networkPolicy
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: [ERROR] disk_iops_read_write and disk_mbps_read_write are only available for UltraSSD disks
        
          on terraform_plugin_test.tf line 11, in resource "azurerm_managed_disk" "test":
          11: resource "azurerm_managed_disk" "test" {
        
--- FAIL: TestAccAzureRMManagedDisk_networkPolicy (104.83s)
FAIL

------- Stderr: -------
2021/04/28 06:39:36 [DEBUG] not using binary driver name, it's no longer needed
2021/04/28 06:39:36 [DEBUG] not using binary driver name, it's no longer needed
```

this is looking good to go!

**Comment by cmendible** [comment]
@katbyte fixed the tests in: d64af57

**Review by katbyte — CHANGES_REQUESTED** [review]
@cmendible - looks like we have another casing issue 😪 

```
------- Stdout: -------
=== RUN   TestAccAzureRMManagedDisk_networkPolicy_create_withAllowPrivate
=== PAUSE TestAccAzureRMManagedDisk_networkPolicy_create_withAllowPrivate
=== CONT  TestAccAzureRMManagedDisk_networkPolicy_create_withAllowPrivate
    testing.go:620: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_managed_disk.test will be updated in-place
          ~ resource "azurerm_managed_disk" "test" {
              ~ disk_access_id        = "/subscriptions/*******/resourceGroups/ACCTESTRG-210428201256204039/providers/Microsoft.Compute/diskAccesses/accda210428201256204039" -> "/subscriptions/*******/resourceGroups/acctestRG-210428201256204039/providers/Microsoft.Compute/diskAccesses/accda210428201256204039"
                id                    = "/subscriptions/*******/resourceGroups/acctestRG-210428201256204039/providers/Microsoft.Compute/disks/acctestd-210428201256204039"
                name                  = "acctestd-210428201256204039"
                tags                  = {
                    "cost-center" = "ops"
                    "environment" = "acctest"
                }
                # (9 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccAzureRMManagedDisk_networkPolicy_create_withAllowPrivate (189.25s)
FAIL

------- Stderr: -------
2021/04/28 20:12:54 [DEBUG] not using binary driver name, it's no longer needed
2021/04/28 20:12:55 [DEBUG] not using binary driver name, it's no longer needed
```

could we open an issue on the SDK describing this issue and link to it on the property? which i believe we'll need to diffsuppress case

**Comment by cmendible** [comment]
Opened issue to the Azure SDK team here: https://github.com/Azure/azure-rest-api-specs/issues/14192

And `suppress.CaseDifference` added to the code here: f0cc2de

**Review by katbyte — APPROVED** [review]
Thanks @cmendible - LGTM 👍

**Pull request merged**